### DISCUSSION TITLE
Slow startup with Spring

### DISCUSSION THREAD

**Discussion by markwimpory** [jOOQ/jOOQ#17782] [Q&A] [answered]
So simply adding jooq to my spring app via maven extends the startup time by up to 60 seconds where nothing is logged.  Thats with no jooq code or anything, just including the jar file.  Ive even tried excluding it from the component scan in case it was doing stuff in the background but no improvement.  Does anyone know whats happening?  This seems pretty major.

Jooq: 3.19.66
Spring: 3.3.4

2024-12-17 08:31:55 INFO  GraphQLWebAutoConfiguration:150 - Enabling cors filter
2024-12-17 08:32:34 INFO  EndpointLinksResolver:60 - Exposing 4 endpoints beneath base path '/actuator'
2024-12-17 08:32:34 DEBUG ConditionEvaluationReportLogger:72 -

**Comment by markwimpory** [comment]
OK so some further information.  Its in a bit of spring called findDerivativeTypes, and its looping over a lot of derivative types that seem to be all jooq SelectFields.  The array is large some 12000 entries and its being called multiple times.

This seems to be driven by the initialisation of the Dslcontext bean.

**Reply by lukaseder** [reply]
Do you have any evidence that this is a problem caused by jOOQ and not by Spring?

**Reply by markwimpory** [reply]
So its the interaction of JOOQ and Spring.  Its interesting and odd.  All I did was comment in the JOOQ dependency in my maven project, no actual JOOQ code even present.  Then when I run the app or a test (integration) there is a pause as shown above in the log.  What I then did was use the debugger to halt the code during the pause, and I discovered that it seems to be spring wrapping the DSLContext in a proxy.  Some part of that does a load of type mapping which seems to be the slow bit.  So its not JOOQ as such but its the interaction of JOOQ and Spring.

**Reply by markwimpory** [reply]
What I could do is create a vanilla empty Spring project and see if the issue is related to the number of beans present in the context, as this is part of a mature application that also using JPA and GraphQL

**Reply by lukaseder** [reply, marked as answer]
I vaguely recall Spring's instrumentation getting overwhelmed with jOOQ's API in some cases, though I don't understand why it's being done in the first place.

Anyway. I don't think this is jOOQ's fault, much like when IDE autocompletion for Kotlin used to have a hard time with the jOOQ API. Sure, the API is vast, but these things hint at Big-O type problems in Spring (or IDEs). We do our own instrumentation in our tests and can navigate the entire API in milliseconds.

I'd research this more, e.g. using VisualVM or any other profiler, and report it to the relevant Spring project.

**Reply by markwimpory** [reply]
So what I tried to start to look into was to stop Spring wrapping DSLContext in a proxy.  But I cant quite figure out how to control that as yet (or if that will create any side affects, I suspect not).  Needs a bit more research, I think there is a configuration class that might help but its all Spring magic.  If you declare a bean in spring its quite easy to stop the proxy using an annotation, but this isnt the case here, Spring must actually have some dormant JOOQ setting built in.

**Reply by lukaseder** [reply]
Well, I think that it's worth a report to the relevant Spring project...
